### PR TITLE
Add mpcontribs parser

### LINKED ISSUE #131: Add argparser for MPContribs database fetch
Adding a new argument parser for remote database feature

### PR DESCRIPTION
**Pull request by sl5035**
Mpcontribs parser function test

Closes #131
@sl5035 this "closes" statement will automatically close the issue when the PR merges.

### CODE REVIEW

File: pydatarecognition/cif_io.py [hunk 1]
```diff
@@ -222,6 +222,21 @@ def cif_read_ext(cif_file_path, client):
     return None
 
 
+def cif_read_mpcontribs(cif_type: str) -> str:
+    """
+    given the type of cif files, returns 'Not implemented'
```
Comment (sbillinge) [inline pydatarecognition/cif_io.py:227]: The first line is a high level description of the what the function does.  This is not clear to me from what you wrote

If this function is just handling the for loop and not anything else, I would say something like
`returns not implemented warning if --mpcontribs_client is selected at the command line`

I would have it take a bool (i.e., the value of `mpcontribs_client` which is `True` or `False`).  Then we have to decide what it returns. If it returns a string that is the error message then we have to have lots of code on the other side to print that, also what string does it return if `mpcontribs` is `False`?  We want it to print nothing in that case for backwards compatibility.  Probably it should return nothing (which is the same as returning `None` in Python I think.

This is why docstrings are so useful....figuring out what the function will do!

Now the name of the function is better if it makes your code readable.  I think something like `set_client(args.mpcontribs_client)` because the name is more saying what the function does (function names are better as verbs).

This raises another question about what the arg could be.  We could have an arg that is `--client` and it can take values `mpcontribs`, `mpc`, or `fs` and is by default `fs`.  Maybe this iwhat you had in mind and is why you wanted to pass in a string.  In that case, ignore the things I said above.

Anyway, great start.  This is what I mwan about the value of writing docstrings

**Review by sbillinge — COMMENTED** [review]
great start, please see comments

**Review by sbillinge — COMMENTED** [review]
Excellent! let's go with that.   Next write a test.  Do you know where/how to do that?

**Comment by sl5035** [comment]
> Excellent! let's go with that. Next write a test. Do you know where/how to do that?

I asked Dom how to use pytest. I've made some edits to this temporary function to see if pytest works

**Review by sbillinge — COMMENTED** [review]
great job.  Please see my inline comment.

Now we need to test for some edge cases.  What do you want the behavior to be if  `"bad_input"` is passed in for example, or something that is not a string?

File: tests/test_cif_io.py [hunk 2]
```diff
@@ -177,4 +177,19 @@ def mockreturn(*args, **kwargs):
...
+
+mpcontribs_tst = [
+    (['mpc'], 'This function is not yet implemented'),
+    (['fs'], None),
+]
+
+
+@pytest.mark.parametrize("inpt", mpcontribs_tst)
+def test_set_client(inpt):
+    if inpt[0] == 'mpc':
```
[the remark this answers is not in the document]
Comment (sbillinge) [reply tests/test_cif_io.py:189]: I just realized I missed the mark.parameterize line there. I edited it.

[the remark this answers is not in the document]
Comment (sbillinge) [reply tests/test_cif_io.py:189]: if there is more than one input, then "input" could be "inputs" and it can be a list.

[the remark this answers is not in the document]
Comment (sbillinge) [reply tests/test_cif_io.py:189]: to test that the code is written correctly you can make the test pass by putting `assert True`, or set `expected` to `True` and `actual = True` and make sure the test passes.

File: tests/test_cif_io.py [hunk 3]
```diff
@@ -177,4 +177,21 @@ def mockreturn(*args, **kwargs):
...
-# End of file.
+
+mpcontribs_tst = [
+    ('mpc', 'This function is not yet implemented'),
+    ('fs', None),
+    (False, 'mpc_client must be either mpc or fs')
+]
+
+
+@pytest.mark.parametrize("inpt, expected", mpcontribs_tst)
```
Comment (sbillinge) [inline tests/test_cif_io.py:188]: I would change `inpt` to `input`.  Never miss an opportunity to make your code more readable by a human.

File: tests/test_cif_io.py [hunk 4]
```diff
@@ -177,4 +177,21 @@ def mockreturn(*args, **kwargs):
         assert actual == expected
         # assert True
 
-# End of file.
+
+mpcontribs_tst = [
+    ('mpc', 'This function is not yet implemented'),
+    ('fs', None),
+    (False, 'mpc_client must be either mpc or fs')
```
Comment (sbillinge) [inline tests/test_cif_io.py:184]: this only tests if `False` is passed in, which is great but a rather specific case so is not adding much in the way of covering different situations.  

Better coverage is if you pass in a string that is not allowed.  It could be anything, such as `'shiny object'` but to make the tests more readable I would make it `'unknown_string'` or something like that.

Another bad input might be something that is not a string at all, but again you get the most coverage by passing in a general non-string object, but again make the intent visible to the code reader.   There are many ways to do that, but maybe passing in `['non_string_input']`

Hopefully this is helping you to get the idea.

File: tests/test_cif_io.py [hunk 5]
```diff
@@ -177,4 +177,21 @@ def mockreturn(*args, **kwargs):
...
+]
+
```
Comment (sbillinge) [inline tests/test_cif_io.py:186]: remove empty lines here.  At some point soon, please take a moment to read, or reread PEP8  which addresses things like this, but in this case, it is easer to see that this list is inputs to this test function if they are not separated by blank lines.

File: tests/test_cif_io.py [hunk 6]
```diff
@@ -177,4 +177,21 @@ def mockreturn(*args, **kwargs):
...
+
+@pytest.mark.parametrize("inpt, expected", mpcontribs_tst)
+def test_set_client(inpt, expected):
+    result = set_client(inpt)
+
+    # TODO: Implement test cases for unexpected input
```
Comment (sbillinge) [inline tests/test_cif_io.py:192]: These bad inputs won't change the test in this case.  Rather they will change the functionality in the function.  You don't put logic in the test or you break the entire logic because the test is passing because of the logic in the test, but it should pass because of hte logic in the function itself!

There is one exception which is if you want your function to raise an exception in a certain situation.  In that case, you have to write a new test function `test_set_client_bad()` that checks that the exception is correctly raised when it is supposed to, but this won't be needed here.

**Review by sbillinge — COMMENTED** [review]
please see my comments.  Because this "back and forth" makes things very slow, in general it is important to work on multiple things in parallel.  These can be multiple branches on the same repo.  To avoid everything grinding to a halt as you learn the process, I suggest that you also now try opening one or two more PRs on different aspects of the project deliverables and you can be coding on each of them and I can be commenting on each of them, in parallel

File: tests/test_cif_io.py (hunk 3 above)
Comment (sl5035) [reply tests/test_cif_io.py:188]: Changing to input shadows the built-in keyword input so I left it as inpt but I will change it ltr!

File: tests/test_cif_io.py [hunk 7]
```diff
@@ -181,17 +181,12 @@ def mockreturn(*args, **kwargs):
...
-
-    # TODO: Implement test cases for unexpected input
-    # if inpt != 'mpc' and inpt != 'fs':
-    #     assert result == expected
+@pytest.mark.parametrize("input, expected", mpcontribs_tst)
+def test_set_client(input, expected):
+    result = set_client(input)
     assert result == expected
 
 # End of file.
```
Comment (sbillinge) [inline tests/test_cif_io.py:192]: delete the end of file statement or add a line-feed, but I prefer to remove end of file comment left over from before.

**Review by sbillinge — COMMENTED** [review]
@sl5035  great!  

We are ready to write the code now! 

1. run tests, ensure it fails.  The easiest is to do this from pycharm if you are using it
2. update the docstring to reflect any new things that emerged during the test writing
3. write code in the function and keep running the tests until the tests all pass.  If you want to catch the different tests separately you can comment out edge-case tests and get the others to pass and gradually uncomment tests until they all pass, but for a simple job like this, if you think carefully, I think you can code all in on lump and we can close this quickly.

**Comment by sbillinge** [comment]
@sl5035 we will also need a "news" item.  Please look in the wiki, but the summary is here:

1. copy news/TEMPLATE.rst to `<branchname>.rst`
2. edit `<branchname>.rst`

This will be used to automatically create the changelog when we release.

**Comment by sl5035** [comment]
I will be deleting the set_client function for now since I don't see the reason to keep it in our code. Using our argument parser and adding the "client" part will suffice. The users can still choose whether to use mpcontribs as backend or local filesystem as backend. 

We can then run our original code if the program is based on fs and I am writing code for the mpc option. The only way we can keep our set_client function is to run the function based on the parameter given as an argument,
`set_client(args['client'])`
but this seems redundant to me.

Please correct me if I'm wrong!

File: pydatarecognition/main.py [hunk 8]
```diff
@@ -63,109 +65,114 @@ def main(verbose=True):
...
-                ciflog.append(ciffile.name)
-            pcd = cif_read(ciffile)
-            try:
-                user_resampled, target_resampled = xy_resample(user_q, user_int, pcd.q,
-                                             pcd.intensity, x_step=args.get('qgrid_interval'))
-                corr_coeff = correlate(user_resampled[1], target_resampled[1])
-            except (AttributeError, ValueError) as e:
-                skipped_cifs.append((ciffile.name, e))
-                continue
+        if args['client'] == 'fs':
```
Comment (sbillinge) [inline pydatarecognition/main.py:75]: This will work but it is almost certainly not the best way to do this.  What we are after is an interface where the code on the front side of the interface is the same regardless of the client.  It has functions like `find_one`, `find_many`, `update_one`, `update_many` and so on.  When `find_one` is called, the code checks what the client is, if it is `fs` it executes code that "finds one" by opening a file on the `fs`, but if it finds that the client is `mpc` it executes code that connects to mpc and fetches that thing.  This is also extensible, so in the future we could have a mongodb backend and this would work and so on.

If you want to see an example of how to structure this, please take a look at the code in regolith

**Review by sbillinge — COMMENTED** [review]
please see my comment.  If you need to talk, please let me know.
